layabout.py

```diff
@@ -1,3 +1,287 @@
+            self.slack = SlackClient(token=self._token)
+
+        # Use whatever token we've got to attempt to connect (or reconnect).
+        return self.slack.rtm_connect()
+
+    def _reconnect(self, retries: int,
+                   backoff: Callable[[int], float]) -> bool:
+        """
+        Attempt to reconnect to the Slack API.
+
+        Args:
+            retries: The number of retry attempts to make if a connection
+                to Slack if not established or is lost.
+            backoff: The strategy used to determine how
+                long to wait between retries. Must take as input the number of
+                the current retry and output a :obj:`float`.
+
+        Returns:
+            Whether the reconnection succeeded.
+        """
+        # TODO: Should retries start at 0 or 1?
+        for retry in range(retries):
+            if self.connect():
+                return True
+            else:
+                interval = backoff(retry)
+                log.debug("Waiting %.3fs before retrying", interval)
+                time.sleep(interval)
+
+        return False
+
+    def run(self, token: str = None, interval: float = 0.5,
+            retries: int = 4, backoff: Callable[[int], float] = None,
+            until: Callable[[List[dict]], bool] = None) -> None:
+        """
+        Connect to the Slack API and run the event handler loop.
+
+        Args:
+            token: A Slack API token. If absent an attempt will be made to use
+                the environment variable supplied at instantiation.
+            interval: The number of seconds to wait between fetching events
+                from the Slack API.
+            retries: The number of retry attempts to make if a connection to
+                Slack if not established or is lost.
+            backoff: The strategy used to determine how long to wait between
+                retries. Must take as input the number of the current retry and
+                output a :obj:`float`. If absent an exponential backoff
+                strategy will be used.
+            until: The condition used to evaluate whether this method
+                terminates. Must take as input an :obj:`list` of :obj:`dict`
+                representing Slack RTM API events and return a :obj:`bool`. If
+                absent this method will run forever.
+
+        Raises:
+            FailedConnection: If connecting to the Slack API fails.
+        """
+        backoff = backoff or _exponential
+        until = until or _forever
+
+        # The initial connection may use a given token or attempt to use an
+        # environment variable.
+        if not self.connect(token=token):
+            if not self._reconnect(retries=retries, backoff=backoff):
+                raise FailedConnection('Failed to connect to the Slack API')
+
+        # TODO: Should we force callers to handle KeyboardInterrupt on their
+        # own, or should we try to handle it for them? 🤔
```

Comment:
You should let them handle it. If they have any system side effects (files, environment variables, child processes) you'll potentially cause them to leak those resources.

I would personally go so far as to say that if you need to catch a `KeyboardInterrupt` then you should re-raise it after.

Reply:
I appreciate your answer, thanks. This is what I was leaning towards, but now I actually have some solid rationale. :grin: